### ISSUE TITLE
Health checking causing panic

### ISSUE THREAD

**Issue by Dennis-Zhang-SH** [minio/minio-go#1876]
I have some codes that will concurrently remove the object, but it paniced on IsOffline function, here's the stack trace:

```log
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xa8 pc=0x11380b6]

goroutine 9990 [running]:
github.com/minio/minio-go/v7.(*Client).IsOffline(...)
        /root/go/pkg/mod/github.com/minio/minio-go/v7@v7.0.36/api.go:358
github.com/minio/minio-go/v7.(*Client).executeMethod(0x0?, {0x29ba2a0?, 0xc0001bc558?}, {0x2644618?, 0xd52ade?}, {0x0, {0x26783e6, 0x19}, {0xc00232b380, 0x27}, ...})
        /root/go/pkg/mod/github.com/minio/minio-go/v7@v7.0.36/api.go:536 +0x56
github.com/minio/minio-go/v7.(*Client).removeObject(0xc0008f2650?, {0x29ba2a0, 0xc0001bc558}, {0x26783e6, 0x19}, {0xc00232b380, 0x27}, {0x0, 0x0, {0x0, ...}, ...})
        /root/go/pkg/mod/github.com/minio/minio-go/v7@v7.0.36/api-remove.go:175 +0x7a5
github.com/minio/minio-go/v7.(*Client).RemoveObject(0xc0008f27b0?, {0x29ba2a0, 0xc0001bc558}, {0x26783e6, 0x19}, {0xc00232b380, 0x27}, {0x0, 0x0, {0x0, ...}, ...})
        /root/go/pkg/mod/github.com/minio/minio-go/v7@v7.0.36/api-remove.go:139 +0xfc
```

Somehow the `&c.healthStatus` in `atomic.LoadInt32(&c.healthStatus)` moved and becoming a dangling pointer.

**Comment by harshavardhana** [comment]
That can only happen if you have an `nil` -> `client` - this looks like a bug in your code.

**Comment by Dennis-Zhang-SH** [comment]
> That can only happen if you have an `nil` -> `client` - this looks like a bug in your code.

what do you mean by `nil -> client`? please be more specific, I call the method like this:
```golang 
go func(path string) {
    if err := p.MinioClient.RemoveObject(context.Background(), "xxx", path, minio.RemoveObjectOptions{}); err != nil {
	log.Printf("remove minio object failed, path: %s, err: %s\n", path, err)
    }
}(minioPath)
```

**Comment by jiuker** [comment]
```
package main

import (
	"context"
	"log"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
)

func main() {
	endpoint := "play.min.io"
	accessKeyID := "Q3AM3UQ867SPQQA43P2F"
	secretAccessKey := "zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG"
	useSSL := true

	// Initialize minio client object.
	minioClient, err := minio.New(endpoint, &minio.Options{
		Creds:  credentials.NewStaticV4(accessKeyID, secretAccessKey, ""),
		Secure: useSSL,
	})
	if err != nil {
		log.Fatalln(err)
	}
	if err := minioClient.RemoveObject(context.Background(), "0213123123test", "053c6cc8cf79c2582a02292708a7992a", minio.RemoveObjectOptions{}); err != nil {
		log.Printf("remove minio object failed, path: %s, err: %s\n", "053c6cc8cf79c2582a02292708a7992a", err)
	}
	log.Printf("%#v\n", minioClient) // minioClient is now set up
}
```
require github.com/minio/minio-go/v7 v7.0.36
I think you should found why your code panic first. @Dennis-Zhang-SH

**Comment by harshavardhana** [comment]
> ```go
> go func(path string) {
>     if err := p.MinioClient.RemoveObject(context.Background(), "xxx", path, minio.RemoveObjectOptions{}); err != nil {
> 	log.Printf("remove minio object failed, path: %s, err: %s\n", path, err)
>     }
> }(minioPath)
> ```

The issue is p.MinioClient is not yet initialized - you might not be waiting on the p.MinioClient initialization

**Comment by Dennis-Zhang-SH** [comment]
> No panic here:
> 
> ```go
> package main
> 
> import (
> 	"context"
> 	"log"
> 
> 	"github.com/minio/minio-go/v7"
> 	"github.com/minio/minio-go/v7/pkg/credentials"
> )
> 
> func main() {
> 	endpoint := "play.min.io"
> 	accessKeyID := "Q3AM3UQ867SPQQA43P2F"
> 	secretAccessKey := "zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG"
> 	useSSL := true
> 
> 	// Initialize minio client object.
> 	minioClient, err := minio.New(endpoint, &minio.Options{
> 		Creds:  credentials.NewStaticV4(accessKeyID, secretAccessKey, ""),
> 		Secure: useSSL,
> 	})
> 	if err != nil {
> 		log.Fatalln(err)
> 	}
> 	if err := minioClient.RemoveObject(context.Background(), "0213123123test", "053c6cc8cf79c2582a02292708a7992a", minio.RemoveObjectOptions{}); err != nil {
> 		log.Printf("remove minio object failed, path: %s, err: %s\n", "053c6cc8cf79c2582a02292708a7992a", err)
> 	}
> 	log.Printf("%#v\n", minioClient) // minioClient is now set up
> }
> ```
> 
> require github.com/minio/minio-go/v7 v7.0.36 I think you should found why your code panic first. @Dennis-Zhang-SH

There are source codes and panic stack traces, check them before judging my code, I already point out.

**Comment by Dennis-Zhang-SH** [comment]
> > ```go
> > go func(path string) {
> >     if err := p.MinioClient.RemoveObject(context.Background(), "xxx", path, minio.RemoveObjectOptions{}); err != nil {
> > 	log.Printf("remove minio object failed, path: %s, err: %s\n", path, err)
> >     }
> > }(minioPath)
> > ```
> 
> The issue is p.MinioClient is not yet initialized - you might not be waiting on the p.MinioClient initialization

I initialize the client in synchronous way, there's no way it's not initialized at the beginning, I am not sure if there's sudden network fluctuations but definitely not nil client, besides why there's no non-nil assert at the beginning of the func?

**Comment by jiuker** [comment]
> I initialize the client in synchronous way, there's no way it's not initialized at the beginning, I am not sure if there's sudden network fluctuations but definitely not nil client, besides why there's no non-nil assert at the beginning of the func?

```golang
package main

import (
	"context"
	"log"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
)

func main() {
	endpoint := "play.min.io"
	accessKeyID := "Q3AM3UQ867SPQQA43P2F"
	secretAccessKey := "zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG"
	useSSL := true

	// Initialize minio client object.
	minioClient, err := minio.New(endpoint, &minio.Options{
		Creds:  credentials.NewStaticV4(accessKeyID, secretAccessKey, ""),
		Secure: useSSL,
	})
	if err != nil {
		log.Fatalln(err)
	}

	minioClient = nil

	if err := minioClient.RemoveObject(context.Background(), "0213123123test", "053c6cc8cf79c2582a02292708a7992a", minio.RemoveObjectOptions{}); err != nil {
		log.Printf("remove minio object failed, path: %s, err: %s\n", "053c6cc8cf79c2582a02292708a7992a", err)
	}
	log.Printf("%#v\n", minioClient) // minioClient is now set up
}
```
panic as yours. So I think some wrong with your code.

**Comment by harshavardhana** [comment]
> I initialize the client in synchronous way, there's no way it's not initialized at the beginning, I am not sure if there's sudden network fluctuations but definitely not nil client, besides why there's no non-nil assert at the beginning of the func?

No the issue is p.MinioClient is not initialized, `panic` is an assert of sort here. It simply shows you have to fix your code.

If you can share your code we can help, otherwise you need to figure this out yourself.

**Comment by Dennis-Zhang-SH** [comment]
> > I initialize the client in synchronous way, there's no way it's not initialized at the beginning, I am not sure if there's sudden network fluctuations but definitely not nil client, besides why there's no non-nil assert at the beginning of the func?
> 
> No the issue is p.MinioClient is not initialized, `panic` is an assert of sort here. It simply shows you have to fix your code.
> 
> If you can share your code we can help, otherwise you need to figure this out yourself.

Of course I am talking about p.MinioClient, @jiuker why did you close it? I said it I initialize the client in synchronous way and it works normally, but there are several goroutine paniced logs, just because you don't think it's a bug so you can close it? that's how you work? I don't want to put pressure on open source projects since it's all provided by free work, but this is really disappointing

**Comment by jiuker** [comment]
> Of course I am talking about p.MinioClient, @jiuker why did you close it? I said it I initialize the client in synchronous way and it works normally, but there are several goroutine paniced logs, just because you don't think it's a bug so you can close it? that's how you work? I don't want to put pressure on open source projects since it's all provided by free work, but this is really disappointing

Since we stably reproduce your scenarios, this is one where I don't think it's an issue, and after we checked the code, we agreed that this is an issue with your own usage. We have no place to remove it. You could try checking that the client is not empty before you use it. If it doesn't check out and it's not null or it's still in panic, we'll take care of that.

**Comment by klauspost** [comment]
@Dennis-Zhang-SH We are working on the information you are giving. It is not our task to guess what you are doing, and you are being anything but forthcoming.

Locking this issue. If you believe there is an issue, do the following:

A) Create a *standalone* reproducer we can execute with "go run main.go" to see the error, which reproduces with the *latest* released version.

You are welcome to use our playground as target `url: "https://play.min.io", accessKey: "Q3AM3UQ867SPQQA43P2F", secretKey: "zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG"`

B) Open a new issue.
